cirrus-ci_env/cirrus-ci_env.py

```diff
@@ -0,0 +1,259 @@
+                # unique name based on alias + matrix attributes.  This is
+                # a very complex process that would be insane to attempt to
+                # duplicate.  Instead, simply require a defined 'name'
+                # attribute in every case, throwing an error if not found.
+                raise ValueError(f"Expecting 'name' attribute in"
+                                 f" '{alias_default}_task'"
+                                 f" or matrix definition: {item}"
+                                 f" for task definition: {task}")
+            # default values for the rendered task
+            matrix_task = dict(alias=alias_default, env=task.get("env").copy())
+            matrix_name = item.get("name", name_default)
+
+            # matrix item env. overwrites task env.
+            matrix_task["env"].update(item.get("env", dict()))
+            matrix_task["env"] = self.render_env(matrix_task["env"])
+            matrix_name = self.render_value(matrix_name, matrix_task["env"])
+
+            # Matrix item overides task dict, overrides global defaults.
+            _ = self.get_type_image(item, self.global_type, self.global_image)
+            matrix_type, matrix_image = self.get_type_image(task, *_)
+            self.init_task_type_image(matrix_task, matrix_type, matrix_image)
+            result[matrix_name] = matrix_task
+        return result
+
+    def render_value(self, value: str, env: dict) -> str:
+        """Given a string value and task env dict, safely render references."""
+        tmp_env = env.copy()  # don't mess up the original
+        tmp_env["__value__"] = value
+        return self.format_env(tmp_env, self.global_env)["__value__"]
+
+    def get_type_image(self, item: dict,
+                       default_type: str = None,
+                       default_image: str = None) -> tuple:
+        """Given Cirrus-CI config or task dict., return instance type and image."""
+        # Order is significant, VMs always override containers
+        if "gce_instance" in item:
+            inst_type = "gcevm"
+            inst_image = item["gce_instance"].get("image_name", default_image)
+        elif "osx_instance" in item:
+            inst_type = "osx"
+            inst_image = item["osx_instance"].get("image", default_image)
+        elif "image" in item.get("container", ""):
+            inst_type = "container"
+            inst_image = item["container"].get("image", default_image)
+        elif "dockerfile" in item.get("container", ""):
+            inst_type = "dockerfile"
+            inst_image = item["container"].get("dockerfile", default_image)
+        else:
+            inst_type = None
+            if self.global_type is not None:
+                inst_type = default_type
+            inst_image = None
+            if self.global_image is not None:
+                inst_image = default_image
+        return (inst_type, inst_image)
```

Comment:
I would tend to not use elif here and use n-returns.

```
if criteria:
    return "dockerfile", item...
if next_ criteria:
...
```
This style makes it very clear once criteria is met, you are finished rather than having to read whole method. 
Additionally, return does not require ()

Reply:
Oh I see what you're saying.  Okay I'll give that a try.